internal/werf/werf.go

```diff
@@ -0,0 +1,220 @@
+		absPath = filepath.Clean(dir)
+	}
+
+	// start from it
+	var currentDir string
+	if fsutils.IsDir(absPath) {
+		currentDir = absPath
+	} else {
+		currentDir = filepath.Dir(absPath)
+	}
+
+	for {
+		result := filepath.Join(currentDir, werfFileName)
+		if fsutils.IsFile(result) {
+			return result
+		}
+		currentDir = filepath.Dir(currentDir)
+		if currentDir == "/" {
+			break
+		}
+	}
+
+	return ""
+}
+
+func parseWerfConfigTemplatesDir(rootDir string, tmpl *template.Template) error {
+	templatesDir := filepath.Join(rootDir, defaultWerfConfigTemplatesDirName)
+	if !fsutils.IsDir(templatesDir) {
+		return nil
+	}
+
+	if err := filepath.WalkDir(templatesDir, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		if filepath.Ext(path) == ".tmpl" {
+			data, err := os.ReadFile(path)
+			if err != nil {
+				return err
+			}
+			name := filepath.ToSlash(path[len(templatesDir)+1:])
+			if err := addTemplate(tmpl, name, string(data)); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addTemplate(tmpl *template.Template, templateName, templateContent string) error {
+	extraTemplate := tmpl.New(templateName)
+	_, err := extraTemplate.Parse(templateContent)
+	return err
+}
+
+func executeTemplate(tmpl *template.Template, name string, data any) (string, error) {
+	buf := bytes.NewBuffer(nil)
+	if err := tmpl.ExecuteTemplate(buf, name, data); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func funcMap(tmpl *template.Template) template.FuncMap {
+	funcMap := sprig.TxtFuncMap()
+	delete(funcMap, "expandenv")
+
+	funcMap["fromYaml"] = func(str string) (map[string]any, error) {
+		m := map[string]any{}
+
+		if err := yaml.Unmarshal([]byte(str), &m); err != nil {
+			return nil, err
+		}
+
+		return m, nil
+	}
+	funcMap["include"] = func(name string, data any) (string, error) {
+		return executeTemplate(tmpl, name, data)
+	}
+	funcMap["tpl"] = func(templateContent string, data any) (string, error) {
+		templateName := generateRandomTemplateFuncName()
+		if err := addTemplate(tmpl, templateName, templateContent); err != nil {
+			return "", err
+		}
+
+		return executeTemplate(tmpl, templateName, data)
+	}
+
+	funcMap["env"] = func(value any, args ...string) (string, error) {
+		if len(args) > 1 {
+			return "", fmt.Errorf("more than 1 optional argument prohibited")
+		}
+
+		envVarName := fmt.Sprint(value)
+
+		var fallbackValue *string
+		if len(args) == 1 {
+			fallbackValue = &args[0]
+		}
+
+		envVarValue, envVarFound := os.LookupEnv(envVarName)
+		if !envVarFound {
+			if fallbackValue != nil {
+				return *fallbackValue, nil
+			}
+			return "", nil
+		}
+
+		if envVarValue == "" && fallbackValue != nil {
+			return *fallbackValue, nil
+		}
+
+		return envVarValue, nil
+	}
+
+	funcMap["required"] = func(msg string, val any) (any, error) {
+		if val == nil {
```

Comment:
u do not need casting to check val == ""
```suggestion
		if val == nil {
			return nil, errors.New(msg)
		} else if val == "" {
				return val, errors.New(msg)
			}
```